src/libfuncs/array.rs

```diff
@@ -335,28 +361,33 @@ pub fn build_tuple_from_span<'ctx, 'this>(
             is_shared,
             &[],
             {
+                // if the array is shared we clone the inner data,
+                // as a tuple does not contain a reference counter.
+
                 let region = Region::new();
                 let block = region.append_block(Block::new(&[]));
 
                 match metadata.get::<DupOverridesMeta>() {
                     Some(dup_overrides_meta) if dup_overrides_meta.is_overriden(&info.ty) => {
-                        let src_ptr = array_ptr;
+                        let src_ptr = array_data_start_ptr;
                         let dst_ptr = value;
 
                         let value = block.load(context, location, src_ptr, tuple_ty)?;
+
+                        // as the array data has the same representation as the tuple,
+                        // we can use the tuple override, which is simpler.
                         let values = dup_overrides_meta
                             .invoke_override(context, &block, location, &info.ty, value)?;
                         block.store(context, location, src_ptr, values.0)?;
                         block.store(context, location, dst_ptr, values.1)?;
                     }
-                    _ => block.memcpy(context, location, array_ptr, value, value_size),
+                    _ => block.memcpy(context, location, array_data_start_ptr, value, value_size),
                 }
 
-                // The following unwrap should be unreachable because an array always has a drop
-                // implementation.
+                // drop the original array (decreasing its reference counter)
                 metadata
                     .get::<DropOverridesMeta>()
-                    .unwrap()
+                    .expect("array always has a drop implementation")
```

Comment:
Why not using ´?´ instead? Just to remove the panic.

Reply:
This `unwrap` is unreachable. An array always has a drop implementation because we need to free the memory.
Does the `?` operator work directly? Or do we need to `.map_err`?

Reply:
We would need to use native_assert. I'll change it to avoid panics.

Reply:
Done